sgx/epid-types/src/quote.rs

```diff
@@ -0,0 +1,461 @@
+                .try_into()
+                .expect("Invalid length of QE security version"),
+        );
+        inner.pce_svn = u16::from_le_bytes(
+            src[PCE_SVN_START..PCE_SVN_END]
+                .try_into()
+                .expect("Invalid length of PCE security version"),
+        );
+        inner.xeid = u32::from_le_bytes(
+            src[XEID_START..XEID_END]
+                .try_into()
+                .expect("Invalid length of XEID"),
+        );
+        inner.basename = basename.into();
+        inner.report_body = report_body.into();
+        inner.signature_len = signature_len as u32;
+
+        retval
+            .aligned_mut(SIGNATURE_START, signature_len)
+            .copy_from_slice(&src[SIGNATURE_START..SIGNATURE_START + signature_len]);
+
+        Ok(retval)
+    }
+}
+
+impl IntelLayout for Quote {
+    const X86_64_CSIZE: usize = QUOTE_MIN_SIZE + QUOTE_SIGLEN_MAX;
+
+    fn intel_size(&self) -> usize {
+        QUOTE_MIN_SIZE + self.signature_len() as usize
+    }
+}
+
+impl Serialize for Quote {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let bytes = self.to_x64_vec();
+        serializer.serialize_newtype_struct("Quote", &bytes[..])
+    }
+}
+
+impl ToX64 for Quote {
+    fn to_x64(&self, dest: &mut [u8]) -> Result<usize, usize> {
+        let signature_len = self.signature_len();
+        let required_len = signature_len as usize + QUOTE_MIN_SIZE;
+        let dest_len = dest.len();
+        if dest_len < required_len {
+            return Err(required_len);
+        }
+
+        dest[VERSION_START..VERSION_END].copy_from_slice(&self.version().to_le_bytes());
+        let value: u16 = self.sign_type().into();
+        dest[SIGN_TYPE_START..SIGN_TYPE_END].copy_from_slice(&value.to_le_bytes());
+
+        self.epid_group_id()
+            .to_x64(&mut dest[EPID_GROUP_ID_START..EPID_GROUP_ID_END])
+            .map_err(|_e| required_len)?;
+
+        dest[QE_SVN_START..QE_SVN_END].copy_from_slice(&self.qe_security_version().to_le_bytes());
+        dest[PCE_SVN_START..PCE_SVN_END]
+            .copy_from_slice(&self.pce_security_version().to_le_bytes());
+        dest[XEID_START..XEID_END].copy_from_slice(&self.xeid().to_le_bytes());
+
+        self.basename()
+            .to_x64(&mut dest[BASENAME_START..BASENAME_END])
+            .map_err(|_e| required_len)?;
+        self.report_body()
+            .to_x64(&mut dest[REPORT_BODY_START..REPORT_BODY_END])
+            .map_err(|_e| required_len)?;
+        dest[SIGLEN_START..SIGLEN_END].copy_from_slice(&signature_len.to_le_bytes());
+
+        if let Some(sigslice) = self.signature() {
+            dest[SIGNATURE_START..(SIGNATURE_START + sigslice.len())].copy_from_slice(sigslice);
+        }
+
+        Ok(required_len)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use bincode::{deserialize, serialize};
+
+    const QUOTE: &[u8] = include_bytes!("test/quote_ok.bin");
```

Comment:
Possible other unittests for other quotes - I think we have a few GOOD and hardening needed quotes.

Reply:
Group-out-of-date and hardening-needed are both IAS responses, not (aesmd-provided) quotes, so they will need to be tested in a (future) `mc-sgx-ias-types` crate.

Reply:
My understanding was that those responses are generated based on the quote provided, so my point was that it may be worth storing some of those quotes and adding tests around them at the quote level. We are already storing some of those example quotes in attest/test_certs. But I see your point, and that test path can be executed at the ias-response testing level.